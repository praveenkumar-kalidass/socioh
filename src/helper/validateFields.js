const validateFields = (schema, state, initialState) => {
  const { error } = schema.validate(state, { abortEarly: false });
  if (error?.details) {
    return error.details.reduce(
      (result, item) => {
        if (state[item.context.key]) {
          return {
            ...result,
            [item.context.key]: item.message,
          };
        }
        return {
          ...result,
          [item.context.key]: '',
        };
      },
      {
        name: '',
        email: '',
        password: '',
      },
    );
  }
  return initialState;
};

export default validateFields;
