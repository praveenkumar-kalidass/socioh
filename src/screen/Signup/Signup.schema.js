import Joi from 'joi';

import { TRANSLATION } from '../../constant';

export const SCHEMA = Joi.object({
  name: Joi.string().alphanum().min(3).max(30).required(),
  email: Joi.string().email({ minDomainSegments: 2, tlds: { allow: ['com'] } }),
  password: Joi.string().alphanum().min(6).max(15).required(),
});

export const INITIAL_STATE = {
  name: '',
  email: '',
  password: '',
};

export const FIELDS = [
  {
    key: 'name',
    label: TRANSLATION.YOUR_NAME,
    placeholder: TRANSLATION.NAME,
  },
  {
    key: 'email',
    label: TRANSLATION.YOUR_EMAIL,
    placeholder: TRANSLATION.EMAIL,
  },
  {
    key: 'password',
    label: TRANSLATION.CREATE_PASSWORD,
    placeholder: TRANSLATION.PASSWORD,
  },
];
