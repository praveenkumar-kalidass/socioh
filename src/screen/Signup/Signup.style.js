import styled from 'styled-components/native';
import { Button, CheckBox, Item, Text } from 'native-base';

import { COLOR, FONT, SPACING } from '../../constant';

export const SignupContainer = styled.View`
  flex: 1;
  justify-content: center;
  padding-horizontal: ${SPACING.SPACE_20};
`;

export const InputContainer = styled.View`
  flex: 1;
`;

export const InputLabel = styled(Text)`
  margin-bottom: ${SPACING.SPACE_10};
`;

export const InputGrid = styled(Item)`
  background-color: ${COLOR.BACKGROUND};
  border-radius: ${SPACING.SPACE_5};
`;

export const InputErrorText = styled(Text)`
  color: ${COLOR.RED};
  font-size: ${FONT.SIZE_12};
  margin-top: ${SPACING.SPACE_10};
`;

export const SignupButton = styled(Button)`
  background-color: ${COLOR.SECONDARY};
  flex: 1;
  justify-content: center;
  margin-horizontal: ${SPACING.SPACE_10};
`;

export const SignupButtonText = styled(Text)`
  color: ${COLOR.BLACK};
  text-align: center;
`;

export const StyledCheckbox = styled(CheckBox)`
  margin-right: ${SPACING.SPACE_20};
`;

export const AgreementContainer = styled.View`
  flex: 1;
  flex-direction: row;
`;
