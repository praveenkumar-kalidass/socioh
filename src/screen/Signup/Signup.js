import React, { useState } from 'react';
import { Card, CardItem, Input, Text } from 'native-base';

import { COLOR, TRANSLATION, CONSTANT } from '../../constant';
import wrapPattern from '../../helper/wrapPattern';
import validateFields from '../../helper/validateFields';
import { SCHEMA, INITIAL_STATE, FIELDS } from './Signup.schema';
import {
  SignupContainer,
  SignupButton,
  SignupButtonText,
  InputGrid,
  InputLabel,
  InputErrorText,
  InputContainer,
  StyledCheckbox,
  AgreementContainer,
} from './Signup.style';

const Signup = () => {
  const [fields, setFields] = useState(INITIAL_STATE);
  const [errorMessages, setErrorMessages] = useState(INITIAL_STATE);

  const handleFieldChange = (label, value) => {
    setFields({
      ...fields,
      [label]: value,
    });
  };

  const validateInput = () => {
    setErrorMessages(validateFields(SCHEMA, fields, INITIAL_STATE));
  };

  return (
    <SignupContainer>
      <Card>
        {FIELDS.map((field, index) => (
          <CardItem key={`signup_input_${index}`}>
            <InputContainer>
              <InputLabel>{field.label}</InputLabel>
              <InputGrid regular>
                <Input
                  placeholder={field.placeholder}
                  value={fields[field.key]}
                  onChangeText={(value) => handleFieldChange(field.key, value)}
                  onBlur={() => validateInput()}
                  secureTextEntry={field === CONSTANT.PASSWORD}
                />
              </InputGrid>
              {Boolean(errorMessages[field.key]) && (
                <InputErrorText>{errorMessages[field.key]}</InputErrorText>
              )}
            </InputContainer>
          </CardItem>
        ))}
        <CardItem>
          <AgreementContainer>
            <StyledCheckbox checked color={COLOR.SECONDARY} />
            <Text>{TRANSLATION.TERMS_AND_CONDITIONS}</Text>
          </AgreementContainer>
        </CardItem>
        <CardItem footer>
          <SignupButton rounded large>
            <SignupButtonText>{TRANSLATION.SIGN_UP}</SignupButtonText>
          </SignupButton>
        </CardItem>
      </Card>
    </SignupContainer>
  );
};

export default wrapPattern({ Component: Signup });
