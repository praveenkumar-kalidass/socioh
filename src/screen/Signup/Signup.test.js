import React from 'react';
import { Alert } from 'react-native';
import { render, act, fireEvent } from '@testing-library/react-native';

import validateFields from '../../helper/validateFields';
import Signup from './index';

jest.mock('../../helper/validateFields');

describe('Signup', () => {
  beforeEach(() => {
    Alert.alert = jest.fn();
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should match snapshot', () => {
    const container = render(<Signup />);

    expect(container).toMatchSnapshot();
  });

  it('should not show validation error when signup name is valid', () => {
    validateFields.mockImplementation(() => ({
      name: '',
      email: '',
      password: '',
    }));

    const { getByTestId, queryByTestId } = render(<Signup />);

    act(() => {
      fireEvent.changeText(getByTestId('signup_input_name'), 'Praveen');
    });
    act(() => {
      fireEvent(getByTestId('signup_input_name'), 'onBlur');
    });

    expect(getByTestId('signup_input_name').props.value).toStrictEqual(
      'Praveen',
    );
    expect(queryByTestId('signup_error_name')).toBeNull();
  });

  it('should show validation error when signup name is not valid', () => {
    validateFields.mockImplementation(() => ({
      name: 'name should have minimum 3 characters',
      email: '',
      password: '',
    }));

    const { getByTestId } = render(<Signup />);

    act(() => {
      fireEvent.changeText(getByTestId('signup_input_name'), 'Pr');
    });
    act(() => {
      fireEvent(getByTestId('signup_input_name'), 'onBlur');
    });

    expect(getByTestId('signup_input_name').props.value).toStrictEqual('Pr');
    expect(getByTestId('signup_error_name').props.children).toStrictEqual(
      'name should have minimum 3 characters',
    );
  });

  it('should not complete signup and show error status when terms & conditions is not agreed', () => {
    validateFields.mockImplementation(() => ({
      name: '',
      email: '',
      password: '',
    }));

    const { getByTestId } = render(<Signup />);

    act(() => {
      fireEvent.changeText(getByTestId('signup_input_name'), 'Praveen');
      fireEvent(getByTestId('signup_input_name'), 'onBlur');
      fireEvent.changeText(
        getByTestId('signup_input_email'),
        'praveen@github.com',
      );
      fireEvent(getByTestId('signup_input_email'), 'onBlur');
      fireEvent.changeText(getByTestId('signup_input_password'), '1234567890');
      fireEvent(getByTestId('signup_input_password'), 'onBlur');
    });
    act(() => {
      fireEvent.press(getByTestId('signup_submit'));
    });

    expect(Alert.alert).toHaveBeenCalledTimes(0);
    expect(getByTestId('signup_terms').props.style.borderColor).toStrictEqual(
      '#ff0000',
    );
  });

  it('should complete signup when all field validations are passed', () => {
    validateFields.mockImplementation(() => ({
      name: '',
      email: '',
      password: '',
    }));

    const { getByTestId } = render(<Signup />);

    act(() => {
      fireEvent.changeText(getByTestId('signup_input_name'), 'Praveen');
      fireEvent(getByTestId('signup_input_name'), 'onBlur');
      fireEvent.changeText(
        getByTestId('signup_input_email'),
        'praveen@github.com',
      );
      fireEvent(getByTestId('signup_input_email'), 'onBlur');
      fireEvent.changeText(getByTestId('signup_input_password'), '1234567890');
      fireEvent(getByTestId('signup_input_password'), 'onBlur');
      fireEvent(getByTestId('signup_terms'), 'onPress');
    });
    act(() => {
      fireEvent.press(getByTestId('signup_submit'));
    });

    expect(Alert.alert).toHaveBeenCalledTimes(1);
  });
});
