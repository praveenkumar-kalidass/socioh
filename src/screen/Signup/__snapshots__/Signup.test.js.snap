// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Signup should match snapshot 1`] = `
Array [
  <View
    style={
      Array [
        Object {
          "position": "absolute",
        },
      ]
    }
  >
    <RNSVGSvgView
      bbHeight={1334}
      bbWidth={750}
      focusable={false}
      height={1334}
      style={
        Array [
          Object {
            "backgroundColor": "transparent",
            "borderWidth": 0,
          },
          Object {
            "flex": 0,
            "height": 1334,
            "width": 750,
          },
        ]
      }
      width={750}
    >
      <RNSVGGroup>
        <RNSVGCircle
          cx={83.75}
          cy={333.5}
          fill={4279943103}
          propList={
            Array [
              "fill",
            ]
          }
          r={93.75}
        />
        <RNSVGCircle
          cx={666.25}
          cy={1187.25}
          fill={4294892136}
          propList={
            Array [
              "fill",
            ]
          }
          r={166.75}
        />
      </RNSVGGroup>
    </RNSVGSvgView>
  </View>,
  <View
    style={
      Array [
        Object {
          "flexBasis": 0,
          "flexGrow": 1,
          "flexShrink": 1,
          "justifyContent": "center",
          "paddingHorizontal": 20,
        },
      ]
    }
  >
    <View
      style={
        Object {
          "backgroundColor": "#fff",
          "borderBottomWidth": 0.5,
          "borderColor": "#ccc",
          "borderLeftWidth": 0.5,
          "borderRadius": 2,
          "borderRightWidth": 0.5,
          "borderTopWidth": 0.5,
          "elevation": 3,
          "flexWrap": "nowrap",
          "marginBottom": 5,
          "marginLeft": 2,
          "marginRight": 2,
          "marginTop": 5,
          "shadowColor": "#000",
          "shadowOffset": Object {
            "height": 2,
            "width": 0,
          },
          "shadowOpacity": 0.1,
          "shadowRadius": 1.5,
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#fff",
            "borderRadius": 2,
            "flexDirection": "row",
            "paddingBottom": 10,
            "paddingLeft": 15,
            "paddingRight": 15,
            "paddingTop": 10,
          }
        }
      >
        <View
          style={
            Array [
              Object {
                "flexBasis": 0,
                "flexGrow": 1,
                "flexShrink": 1,
              },
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#000",
                  "fontFamily": "System",
                  "fontSize": 16,
                },
                Object {
                  "marginBottom": 10,
                },
              ]
            }
          >
            Your name
          </Text>
          <View
            accessible={true}
            focusable={false}
            forwardedRef={[Function]}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "#f0f8f8",
                "borderBottomWidth": 1,
                "borderColor": "#D9D5DC",
                "borderLeftWidth": 1,
                "borderRadius": 5,
                "borderRightWidth": 1,
                "borderTopWidth": 1,
                "borderWidth": undefined,
                "flexDirection": "row",
                "marginLeft": 2,
                "opacity": 1,
              }
            }
          >
            <TextInput
              allowFontScaling={true}
              editable={true}
              onBlur={[Function]}
              onChangeText={[Function]}
              placeholder="Name"
              placeholderTextColor="#575757"
              rejectResponderTermination={true}
              secureTextEntry={false}
              style={
                Object {
                  "color": "#000",
                  "flex": 1,
                  "fontSize": 17,
                  "height": 50,
                  "paddingLeft": 8,
                  "paddingRight": 5,
                  "top": 1.5,
                }
              }
              underlineColorAndroid="rgba(0,0,0,0)"
              value=""
            />
          </View>
        </View>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#fff",
            "borderRadius": 2,
            "flexDirection": "row",
            "paddingBottom": 10,
            "paddingLeft": 15,
            "paddingRight": 15,
            "paddingTop": 10,
          }
        }
      >
        <View
          style={
            Array [
              Object {
                "flexBasis": 0,
                "flexGrow": 1,
                "flexShrink": 1,
              },
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#000",
                  "fontFamily": "System",
                  "fontSize": 16,
                },
                Object {
                  "marginBottom": 10,
                },
              ]
            }
          >
            Your e-mail
          </Text>
          <View
            accessible={true}
            focusable={false}
            forwardedRef={[Function]}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "#f0f8f8",
                "borderBottomWidth": 1,
                "borderColor": "#D9D5DC",
                "borderLeftWidth": 1,
                "borderRadius": 5,
                "borderRightWidth": 1,
                "borderTopWidth": 1,
                "borderWidth": undefined,
                "flexDirection": "row",
                "marginLeft": 2,
                "opacity": 1,
              }
            }
          >
            <TextInput
              allowFontScaling={true}
              editable={true}
              onBlur={[Function]}
              onChangeText={[Function]}
              placeholder="E-mail"
              placeholderTextColor="#575757"
              rejectResponderTermination={true}
              secureTextEntry={false}
              style={
                Object {
                  "color": "#000",
                  "flex": 1,
                  "fontSize": 17,
                  "height": 50,
                  "paddingLeft": 8,
                  "paddingRight": 5,
                  "top": 1.5,
                }
              }
              underlineColorAndroid="rgba(0,0,0,0)"
              value=""
            />
          </View>
        </View>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#fff",
            "borderRadius": 2,
            "flexDirection": "row",
            "paddingBottom": 10,
            "paddingLeft": 15,
            "paddingRight": 15,
            "paddingTop": 10,
          }
        }
      >
        <View
          style={
            Array [
              Object {
                "flexBasis": 0,
                "flexGrow": 1,
                "flexShrink": 1,
              },
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#000",
                  "fontFamily": "System",
                  "fontSize": 16,
                },
                Object {
                  "marginBottom": 10,
                },
              ]
            }
          >
            Create password
          </Text>
          <View
            accessible={true}
            focusable={false}
            forwardedRef={[Function]}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "#f0f8f8",
                "borderBottomWidth": 1,
                "borderColor": "#D9D5DC",
                "borderLeftWidth": 1,
                "borderRadius": 5,
                "borderRightWidth": 1,
                "borderTopWidth": 1,
                "borderWidth": undefined,
                "flexDirection": "row",
                "marginLeft": 2,
                "opacity": 1,
              }
            }
          >
            <TextInput
              allowFontScaling={true}
              editable={true}
              onBlur={[Function]}
              onChangeText={[Function]}
              placeholder="Password"
              placeholderTextColor="#575757"
              rejectResponderTermination={true}
              secureTextEntry={false}
              style={
                Object {
                  "color": "#000",
                  "flex": 1,
                  "fontSize": 17,
                  "height": 50,
                  "paddingLeft": 8,
                  "paddingRight": 5,
                  "top": 1.5,
                }
              }
              underlineColorAndroid="rgba(0,0,0,0)"
              value=""
            />
          </View>
        </View>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#fff",
            "borderRadius": 2,
            "flexDirection": "row",
            "paddingBottom": 10,
            "paddingLeft": 15,
            "paddingRight": 15,
            "paddingTop": 10,
          }
        }
      >
        <View
          style={
            Array [
              Object {
                "flexBasis": 0,
                "flexDirection": "row",
                "flexGrow": 1,
                "flexShrink": 1,
              },
            ]
          }
        >
          <View
            accessible={true}
            focusable={false}
            forwardedRef={[Function]}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "backgroundColor": "#feda68",
                "borderBottomWidth": 1,
                "borderColor": "#feda68",
                "borderLeftWidth": 1,
                "borderRadius": 13,
                "borderRightWidth": 1,
                "borderTopWidth": 1,
                "height": 20,
                "left": 10,
                "marginRight": 20,
                "opacity": 1,
                "overflow": "hidden",
                "paddingBottom": 0,
                "paddingLeft": 3,
                "width": 20,
              }
            }
          >
            <Text
              allowFontScaling={false}
              style={
                Array [
                  Object {
                    "color": undefined,
                    "fontSize": 12,
                  },
                  Object {
                    "color": "#fff",
                    "fontSize": 25.555555555555554,
                    "lineHeight": 21,
                    "marginTop": undefined,
                    "textShadowRadius": 0,
                  },
                  Object {
                    "fontFamily": "Ionicons",
                    "fontStyle": "normal",
                    "fontWeight": "normal",
                  },
                  Object {},
                ]
              }
            >
              
            </Text>
          </View>
          <Text
            style={
              Object {
                "color": "#000",
                "fontFamily": "System",
                "fontSize": 16,
              }
            }
          >
            I agree with terms & conditions
          </Text>
        </View>
      </View>
      <View
        footer={true}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#fff",
            "borderBottomWidth": null,
            "borderRadius": 2,
            "flexDirection": "row",
            "paddingBottom": 10,
            "paddingLeft": 15,
            "paddingRight": 15,
            "paddingTop": 10,
          }
        }
      >
        <View
          accessible={true}
          focusable={false}
          forwardedRef={[Function]}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "alignSelf": "flex-start",
              "backgroundColor": "#feda68",
              "borderBottomWidth": null,
              "borderColor": "#007aff",
              "borderLeftWidth": null,
              "borderRadius": 57,
              "borderRightWidth": null,
              "borderTopWidth": null,
              "borderWidth": undefined,
              "elevation": 2,
              "flexBasis": 0,
              "flexDirection": "row",
              "flexGrow": 1,
              "flexShrink": 1,
              "height": 60,
              "justifyContent": "center",
              "marginHorizontal": 10,
              "opacity": 1,
              "paddingBottom": 6,
              "paddingTop": 6,
              "shadowColor": undefined,
              "shadowOffset": undefined,
              "shadowOpacity": undefined,
              "shadowRadius": undefined,
            }
          }
        >
          <Text
            style={
              Array [
                Object {
                  "backgroundColor": "transparent",
                  "color": "#fff",
                  "fontFamily": "System",
                  "fontSize": 22,
                  "marginLeft": 0,
                  "marginRight": 0,
                  "paddingLeft": 16,
                  "paddingRight": 16,
                },
                Object {
                  "color": "#000000",
                  "textAlign": "center",
                },
              ]
            }
          >
            Sign up
          </Text>
        </View>
      </View>
    </View>
  </View>,
]
`;
